src/main/java/org/truffleruby/language/methods/AddMethodNode.java

```diff
@@ -44,18 +46,38 @@ public void addMethod(DynamicObject module, InternalMethod method, Visibility vi
             visibility = Visibility.PRIVATE;
         }
 
-        method = method.withVisibility(visibility);
+        if (Layouts.MODULE.getFields(module).isRefinement()) {
+            final DynamicObject refinedClass = Layouts.MODULE.getFields(module).getRefinedClass();
+            addRefinedMethodEntry(refinedClass, method, visibility);
+        }
+
+        doAddMethod(module, method, visibility);
+    }
 
+    @TruffleBoundary
+    private void addRefinedMethodEntry(DynamicObject module, InternalMethod method, Visibility visibility) {
+        final MethodLookupResult result = ModuleOperations.lookupMethodCached(module, method.getName(), null);
+        final InternalMethod originalMethod = result.getMethod();
+        if (originalMethod == null) {
+            doAddMethod(module, method.withRefined(true).withOriginalMethod(null), visibility);
+        } else if (originalMethod.isRefined()) {
+            // Already marked as refined
```

Comment:
Maybe an edge case: I could imagine a scenario where the same method is refined twice so the original method maybe incorrect then.  Could handle this in a later PR though.